dwarf/main.c

```diff
@@ -0,0 +1,1011 @@
+					}
+					strng_free (typename);
+				}
+				break;
+			case DW_TAG_pointer_type:
+				{
+				    strng *t_format = strng_new ("p");
+					strngcat (format, *format, *t_format, "");
+					strng_free (t_format);
+				}
+				break;
+			case DW_TAG_structure_type:
+				{
+					strng *t_format = strng_new ("");
+					
+					strng_free (t_format);
+				}
+				break;
+			default:
+			    //TODO: not implemented yet.
+				printf("Lol!! WTF are you doing?? Get your ass back here and fix my implementation");
+				exit(1);
+			}
+
+			char *name = 0;
+			strng *namestring;
+			res = dwarf_diename (child, &name, &error);
+			if (res == DW_DLV_ERROR) {
+				exit(1);
+			}
+			if (res == DW_DLV_NO_ENTRY) {
+				exit(1);
+			}
+			namestring = strng_new (name);
+			strngcat (nm, *nm, *namestring, " ");
+			strng_free (namestring);
+			dwarf_dealloc (dbg, name, DW_DLA_STRING);
+
+			res = dwarf_siblingof (dbg, child, &sib_die, &error);
+			if (res == DW_DLV_ERROR) {
+				exit(1);
+			}
+
+			if (res == DW_DLV_NO_ENTRY) {
+				break;
+			}
+
+			dwarf_dealloc (dbg, child, DW_DLA_DIE);
+			child = sib_die;
+		}
+	}
+
+	strngcat (jsonoutput, *jsonoutput, *out, "");
+	strngcat (jsonoutput, *jsonoutput, *format, " ");
+	strngcat (jsonoutput, *jsonoutput, *nm, " ");
+	return 0;
+}
+#endif
+
+//The following function was supposed to be used with print_struct_type
+static Dwarf_Half get_type_tag (Dwarf_Debug dbg, Dwarf_Die die) {
+	int res = DW_DLV_ERROR;
+	Dwarf_Half tag = 0;
+	Dwarf_Error error = 0;
+	Dwarf_Die typedie = 0;
+	Dwarf_Attribute attr = 0;
+	Dwarf_Off offset = 0;
+
+	res = dwarf_attr (die, DW_AT_type, &attr, &error);
+	if (res == DW_DLV_ERROR) {
+		res = -1;
+		printf ("error");
+		goto ret;
+	}
+	if (res == DW_DLV_NO_ENTRY) {
+		res = -1;
+		printf ("error");
+		goto ret;
+	}
+
+	res = dwarf_global_formref (attr, &offset, &error);
+	if (res == DW_DLV_ERROR) {
+		res = -1;
+		printf ("error");
+		goto ret;
+	}
+
+	res = dwarf_offdie (dbg, offset, &typedie, &error);
+	if (res != DW_DLV_OK) {
+		res = -1;
+		printf ("error");
+		goto ret;
+	}
+
+	res = dwarf_tag (typedie, &tag, &error);
+	if (res != DW_DLV_OK) {
+		res = -1;
+		printf ("error");
+		goto ret;
+	}
+	res = tag;
+
+ ret:
+	dwarf_dealloc (dbg, attr, DW_DLA_ATTR);
+	dwarf_dealloc (dbg, error, DW_DLA_ERROR);
+	dwarf_dealloc (dbg, typedie, DW_DLA_DIE);
+	return res;
+}
+
+static int get_number (Dwarf_Attribute attr, Dwarf_Unsigned *val) {
+	int res = DW_DLV_ERROR;
+	Dwarf_Signed sval = 0;
+	Dwarf_Unsigned uval = 0;
+	Dwarf_Error error = 0;
+	res = dwarf_formudata (attr, &uval, &error);
+	if (res == DW_DLV_OK) {
+		*val = uval;
+		return 0;
+	}
+	res = dwarf_formsdata (attr, &sval, &error);
+	if (res == DW_DLV_OK) {
+		*val = sval;
+		return 0;
+	}
+	return -1;
+}
+
+static int get_array_length(Dwarf_Debug dbg, Dwarf_Die die, unsigned int *len) {
+	Dwarf_Attribute attr = 0;
+	Dwarf_Error error = 0;
+	Dwarf_Half tag = 0;
+	Dwarf_Unsigned bound = 0;
+	int res = DW_DLV_ERROR;
+
```

Comment:
it is just to be on safe side to have error if the return value is not set. I can change it if it actually doesn't matter.